Content and planning of package management for Debian and HPC (#6)
Simulation-Software-Engineering/Lecture-Material: building-and-packaging/material/packaging_debian_notes.md

# Debian packages

## Learning goals

- How to make a CMake project installable.
- How to package a CMake project via `CPack`.
- How to create a Debian (`deb`) package of your program/library using CPack.
- How to check the resulting Debian package.

## Introduction

- Now:
    - Debian packages for C++ via `CMake`
    - HPC suitable packages for Python and C++
        - This requires a well-defined installation routine.
- Rough plan
    1. Extend CMake configuration from previous lecture to have an `install` target as well as proper prefix and installation layout (`/bin`, `/include`, `/lib`...)
    2. Extend CMake configuration to contain a CPack section (rather minimal) such that we can create a package (`tar.gz` file)
    3. Extend the CMake configuration to create a Debian package

>     4. Check the Debian package with `lintian` and potentially improve the package

Author: It moved on the slides now.

## 1. Preparing Project

- Make sure two files are created (a library and en executable). This should be the case when starting with the code from last week's lecture.
- We add the following changes
    - Adding a version number to the project. This is not needed yet, but I am afraid of forgetting it later.

>     - We mark the include files needed to use the `libsse` library, i.e., `sse/sse.hpp` publicly visible.

Reviewer: Why is this necessary?

Author: Why is what necessary? Creation of a library and executable or the marking of the headers?

- Regarding the library and executable I have commented on the PR already. 
- Regarding the headers we need to tell CMake what are important headers to be used by external projects. By default all headers are assumed to be used only within the project.

Reviewer: I meant the "mark as public".

Author: I find [the documentation on this](https://cmake.org/cmake/help/latest/prop_tgt/PUBLIC_HEADER.html) also a bit confusing. It is mentioned there together with some `FRAMEWORK` property which seems to be only relevant for some Apple OSes. However, if I do not set this property in my `CMakeLists.txt`, the include files are not installed.

Author: I add an additional remark to the notes and I will try to mention this also in the lecture.

    - We define the target's include directory, i.e., where do we have to look for the dependency depending on the type of build (internally, top-level cmake project, external dependency)

>     - We want to make sure that we have an `install` target for make that, at the same time, considers typical installation directories. That means `make install` should binaries in a `<prefix>/bin`, libraries in `<prefix>/lib` or `<prefix>/lib/helloworld`, include files in `<prefix>/include` etc. We include the predefined CMake macro `GNUInstallDirs` to get predefined variables like the installation directories.

Reviewer: ```suggestion
    - We want to make sure that we have an `install` target for make that, at the same time, considers typical installation directories. That means `make install` should install binaries in a `<prefix>/bin`, libraries in `<prefix>/lib` or `<prefix>/lib/helloworld`, include files in `<prefix>/include` etc. We include the predefined CMake macro `GNUInstallDirs` to get predefined variables such as the installation directories.
```

Changes:

```diff
- project("HelloWorld")
+ project("HelloWorld" VERSION 0.1.0)
+ add_library(sse sse/sse.cpp)
+
+ set_target_properties(sse PROPERTIES PUBLIC_HEADER sse/sse.hpp)

+target_include_directories(sse
+    PRIVATE
+        # where the library itself will look for its internal headers
+        ${CMAKE_CURRENT_SOURCE_DIR}/helloworld
+    PUBLIC
+        # where top-level project will look for the library's public headers
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/helloworld>
+        # where external projects will look for the library's public headers
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/helloworld>
+)

+# Create install targets
+include(GNUInstallDirs)
+install(TARGETS helloworld sse
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+  PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/helloworld
+  INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/helloworld
+  )
```

> - Fire up a prepared containerf or testing. Container is beneficial as it allows us to mess around with the container's paths without messing up our own system. From the example repository I run

Reviewer: ```suggestion
- Fire up a prepared container for testing. Container is beneficial as it allows us to mess around with the container's paths without messing up our own system. From the example repository I run
```

  ```bash
  sudo docker run --rm -it --name ubuntu-packaging --mount type=bind,source="$(pwd)",target=/mnt/cpack-example jaustar/debian-package-building-base
  ```

- Build the library and install it

  ```bash
  mkdir build && cd build
  cmake -DBUILD_SHARED_LIBS=ON ..
  make -j
  sudo make install
  ```

> - Installs in `prefix=/usr/local`, i.e., it sets `CMAKE_INSTALL_PREFIX`.

Reviewer: Who sets this? That's the default value of CMake? Or does CPack set this variable?

Author: I rephrased that. CMake uses `/usr/local` as default value for `CMAKE_INSTALL_PREFIX`

  ```bash
  -- Install configuration: ""
  -- Installing: /usr/local/bin/helloworld
  -- Installing: /usr/local/lib/libsse.so
  -- Installing: /usr/local/include/helloworld/sse.hpp
  ```

>   **Note** that we have an empty configuration here. It would make sense to build the project with a proper configuration (`Debug`, `Release`...) by setting `CMAKE_BUILD_TYPE`.

Reviewer: Why not doing it? Students know this option from last week.

Author: Mainly because it slipped through. I was also not sure whether I should show it first without or with the empty configuration. I will try to remember to configure everything with `-DCMAKE_BUILD_TYPE=Release` during the lecture and mention that one should always specify some build type.

    - Full list of statements in [documentation](https://cmake.org/cmake/help/latest/module/CPack.html). Note, that the current statements are common properties of the packaging module.

  ```diff
  + set(CPACK_PACKAGE_NAME ${PROJECT_NAME})
  +
  + set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "SSE hello world example project"
  +   CACHE STRING "Extended summary.")
  +
  + set(CPACK_PACKAGE_VENDOR "SSE Lecturers / Employer")
  +
  + set(CPACK_PACKAGE_CONTACT "firstname.lastname@example.com")
  + set(CPACK_PACKAGE_MAINTAINERS "SSE lecturers ${CPACK_PACKAGE_CONTACT}")
  +
  + set(CPACK_PACKAGE_HOMEPAGE_URL "https://simulation-software-engineering.github.io/homepage/")
  +
  + set(CPACK_PACKAGE_VERSION_MAJOR ${PROJECT_VERSION_MAJOR})
  + set(CPACK_PACKAGE_VERSION_MINOR ${PROJECT_VERSION_MINOR})
  + set(CPACK_PACKAGE_VERSION_PATCH ${PROJECT_VERSION_PATCH})
  +
  + include(CPack)
  ```

- Start up container again and then prepare project

  ```bash
  mkdir build && cd build
  cmake -DBUILD_SHARED_LIBS=ON ..
  ```

- This will create new files `CPackConfig.cmake` and `CPackSourceConfig.cmake`
- We can create a package, e.g., `tar.gz`

  ```bash
  cpack -G TGZ

>   ```

Reviewer: Wouldn't `make package` do the same thing?
That's how I understood it. What did I get the wrong way?

Author: `make package` will create the default package(s) specified in the CPack configuration. In our case it will create three or for packages (including a `tar.gz`) package with are predefined in our case since we did not set [the `CPACK_GENERATOR` variable](https://cmake.org/cmake/help/latest/module/CPack.html#variable:CPACK_GENERATOR). `cpack` gives much more fine-grained control over what package to create.

Reviewer: OK, so, if we set the correct options in the CMake file, `make package` does the trick, right?
Will you show this as well? Or too much?

Author: I don't have it on my list at the moment, but if you would prefer this, there is no issue adding that.

What are you using for packaging preCICE? I had the feeling that `cpack` shows up more in tutorials. I must admit that I wasn't really aware of the `package` target.
